### PR TITLE
[fix] torch.kthvalue : handle non-contiguous CUDA tensor 

### LINKED ISSUE #45721: kthvalue incorrect with strided GPU tensor
## 🐛 Bug

When input tensor is on GPU and is strided, `kthvalue` will give wrong values and indices.

## To Reproduce

```
>>> import torch
>>> a = torch.tensor([0, -1, 1, -2, 2]).cuda()[::2]
>>> a
tensor([0, 1, 2], device='cuda:0')
>>> a.kthvalue(2)
torch.return_types.kthvalue(
values=tensor(0, device='cuda:0'),
indices=tensor(0, device='cuda:0'))
```

## Expected behavior

The behavior on CPU as below is expected:
```
>>> import torch
>>> a = torch.tensor([0, -1, 1, -2, 2])[::2]
>>> a
tensor([0, 1, 2])
>>> a.kthvalue(2)
torch.return_types.kthvalue(
values=tensor(1),
indices=tensor(1))
```
It seems that `kthvalue` is performed on `[0, -1, 1]`, the tensor with the same start address and size as `a` but stride 1.
A workaround is to use `a.contiguous().kthvalue(2)`, but this is not the expected way to use `kthvalue`.

## Environment

```
Collecting environment information...
PyTorch version: 1.7.0.dev20201001
Is debug build: True
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.6 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.2.89
GPU models and configuration: 
GPU 0: GeForce GTX 1080 Ti
GPU 1: GeForce GTX 1080 Ti

Nvidia driver version: 440.64
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.7.0.dev20201001
[pip3] torchvision==0.8.0.dev20201001
[conda] Could not collect
```

cc @ezyang @gchanan @zou3519 @ngimel

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #45721 

TODO
* [x] Test

### CODE REVIEW

File: aten/src/ATen/native/cuda/SortingKthValue.cu [hunk 1]
```diff
@@ -241,7 +241,7 @@ std::tuple<Tensor&, Tensor&> kthvalue_out_cuda(
     bool keepdim) {
   auto result = [&]() {
     NoNamesGuard guard;
-    return kthvalue_out_impl_cuda(values, indices, self, k, dim, keepdim);
+    return kthvalue_out_impl_cuda(values, indices, self.contiguous(), k, dim, keepdim);
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/SortingKthValue.cu:244]: Would you just add a comment here or above the kthvalue_out_impl_cuda() that the function expects (and only works properly on) contiguous tensors?

File: test/test_torch.py [hunk 2]
```diff
@@ -9010,6 +9010,13 @@ def test_kthvalue(self, device, dtype):
         self.assertEqual(res1val, res2val, atol=0, rtol=0)
         self.assertEqual(res1ind, res2ind, atol=0, rtol=0)
 
+        # non-contiguous [Reference: https://github.com/pytorch/pytorch/issues/45721]
+        non_contig_t = torch.tensor([0, -1, 1, -2, 2], dtype=dtype, device=device)[::2]
+        expected_val, expected_ind = non_contig_t.contiguous().kthvalue(2)
```
Comment (mruberry) [inline test/test_torch.py:9015]: Would you also compare it with the cpu?

```
expected_val_cpu, expected_ind_cpu = non_contig_t.kthvalue(2)
...
self.assertEqual(expected_val_cpu, out_val.cpu(), atol=0, rtol=0)
self.assertEqual(expected_ind_cpu, out_ind.cpu(), atol=0, rtol=0)
```

**Review by mruberry — APPROVED** [review]
Hey @kshitij12345 Thanks for fixing this high priority issue! This is a great, simple fix. Maybe in the future we should consider if the kth value kernel should be updated to handle discontiguous inputs, too.

I just made two small suggestions, and it looks like this PR needs to rebased, too (sorry about that). Just ping me when they're done!

Since this fixes a silent correctness issue I think we'll want to put it in the 1.7 release, too.

**Comment by kshitij12345** [comment]
@mruberry 

Have addressed the comments. PTAL:)

**Pull request closed without merging**